File: src/lib/holidays.ts
import { readHolidays } from "@/lib/tauri";

// Кешируем загрузку: holidays.json не меняется в течение сессии (обновится
// через manifest при следующем старте), а CalculatorField монтируется на
// каждой форме — без кеша били бы Rust-команду по разу на каждый калькулятор.
let cache: ReadonlySet<string> | null = null;
let pending: Promise<ReadonlySet<string>> | null = null;

/// Возвращает набор нерабочих праздничных дней (YYYY-MM-DD). Пустой Set, если
/// файл недоступен — калькуляторы в этом случае переносят только сб/вс.
export async function getHolidays(): Promise<ReadonlySet<string>> {
  if (cache) return cache;
  if (pending) return pending;
  pending = readHolidays()
    .then((list) => {
      cache = new Set(list);
      return cache;
    })
    .catch(() => {
      cache = new Set();
      return cache;
    })
    .finally(() => {
      pending = null;
    });
  return pending;
}

File: src-tauri/src/commands/holidays.rs
use std::path::{Path, PathBuf};

fn exe_dir() -> Result<PathBuf, String> {
    let exe = std::env::current_exe().map_err(|e| e.to_string())?;
    exe.parent()
        .map(Path::to_path_buf)
        .ok_or_else(|| "executable has no parent".to_string())
}

/// Читает производственный календарь — массив строк YYYY-MM-DD из
/// `dictionaries/holidays.json` рядом с .exe. На отсутствии/ошибке парсинга
/// возвращает пустой вектор: фронт работает в режиме «только сб/вс»
/// (см. `src/lib/date-utils.ts::isNonBusinessDay`).
#[tauri::command]
pub fn read_holidays() -> Vec<String> {
    let path = match exe_dir() {
        Ok(p) => p.join("dictionaries").join("holidays.json"),
        Err(_) => return Vec::new(),
    };
    let raw = match std::fs::read_to_string(&path) {
        Ok(s) => s,
        Err(_) => return Vec::new(),
    };
    match serde_json::from_str::<Vec<String>>(&raw) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("[read_holidays] parse {}: {e}", path.display());
            Vec::new()
        }
    }
}
